quickwit-storage/src/cache/disk/access_log.rs

```diff
@@ -0,0 +1,570 @@
+        free_slots,
+        entries,
+        key_mapping,
+    })
+}
+
+#[derive(Copy, Clone)]
+pub(crate) enum Event {
+    ReadAccess { file_key: u64, accessed_at: u64 },
+    Write { entry: FileEntry },
+    Remove { file_key: u64 },
+    Shutdown,
+}
+
+#[derive(Clone)]
+pub(crate) struct AccessLogWriter(Sender<Event>);
+
+impl AccessLogWriter {
+    fn send_event(&self, event: Event) {
+        // TODO: Should this be handled?
+        let _ = self.0.send(event);
+    }
+
+    pub(crate) fn register_read(&self, file_key: u64, accessed_at: Duration) {
+        self.send_event(Event::ReadAccess {
+            file_key,
+            accessed_at: accessed_at.as_secs(),
+        })
+    }
+
+    pub(crate) fn register_write(&self, entry: FileEntry) {
+        self.send_event(Event::Write { entry })
+    }
+
+    pub(crate) fn register_removal(&self, file_key: u64) {
+        self.send_event(Event::Remove { file_key })
+    }
+}
+
+impl Drop for AccessLogWriter {
+    fn drop(&mut self) {
+        let _ = self.0.send(Event::Shutdown);
+    }
+}
+
+fn start_background_task(changes: Receiver<Event>, data: LoadedData, writer: File) {
+    std::thread::Builder::new()
+        .name("disk-cache-background-worker".to_string())
+        .spawn(move || {
+            let instance = FileAccessTask {
+                pending_changes: changes,
+                free_slots: data.free_slots,
+                entries: data.entries,
+                key_mapping: data.key_mapping,
+                writer,
+            };
+
+            instance.start()
+        })
+        .expect("spawn background thread"); // TODO: Handle error.
+}
+
+struct FileAccessTask {
+    pending_changes: Receiver<Event>,
+
+    /// A queue of free slots that can be overwritten
+    /// without needing to extend the file.
+    free_slots: Vec<usize>,
+
+    /// A set of file entry rows.
+    /// This essentially mirrors the layout of rows on disk including their order.
+    entries: Vec<FileEntry>,
+
+    /// A lookup dictionary to map a given file key to their relevant index where
+    /// the file entry lies within `entries`.
+    key_mapping: HashMap<FileKey, usize>,
+
+    writer: File,
+}
+
+impl FileAccessTask {
+    fn start(mut self) {
+        loop {
+            // Attempt to pop events off of the backlog.
+            while let Ok(event) = self.pending_changes.try_recv() {
+                if let Event::Shutdown = event {
+                    info!("Cache Shutdown: Received terminating event. Flushing buffers.");
+                    self.try_flush_data();
+                    return;
+                }
+
+                if let Err(e) = self.handle_event(event) {
+                    error!(error = ?e, "IO Cache Error: Failed to handle event due to the the given IO error.");
+                }
+            }
+
+            self.try_flush_data();
+
+            // Wait until the next set of events come through.
+            match self.pending_changes.recv() {
```

Comment:
is there a reason why you did not put this match statement at the top of the loop?